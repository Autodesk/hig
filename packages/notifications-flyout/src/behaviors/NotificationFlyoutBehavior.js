/* eslint-disable react/no-unused-prop-types */

import { Component } from "react";
import PropTypes from "prop-types"; 

import parseNotifications from "./parseNotifications";

/** @typedef {import("./parseNotifications").Input} NotificationsInput */
/** @typedef {import("./parseNotifications").ParsedNotification} ParsedNotification */
/** @typedef {import("./NotificationBehavior").Payload} NotificationBehaviorPayload */

/**
 * @typedef {Object} Payload
 * @property {Function} dismissNotification
 * @property {ParsedNotification[]} notifications
 * @property {boolean} showUnreadCount
 * @property {number} unreadCount
 */

/**
 * @typedef {Object} Props
 * @property {function(Payload): import("react").ReactElement} [children]
 * @property {NotificationsInput} [notifications]
 * @property {number} [unreadCount]
 */

/**
 * @typedef {Object} State
 * @property {string[]} dismissedNotifications An array of notification IDs that have been dismissed
 * @property {ParsedNotification[]} notifications
 * @property {string[]} readNotifications An array of notification IDs that have been read
 */

export default class NotificationFlyoutBehavior extends Component {
  /** @type {Props} */
  props;

  static propTypes = {
    /** Render prop */
    children: PropTypes.func.isRequired,
    /** Rendered notifications */
    notifications: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.node, PropTypes.shape({})])
      )
    ]),
    /** When provided, it overrides the derived unread notification count */
    unreadCount: PropTypes.number
  };

  /** @type {State} */
  state = {
    dismissedNotifications: [],
    notifications: [],
    readNotifications: []
  };

  /**
   * @param {Props} nextProps
   * @returns {State | null}
   */
  static getDerivedStateFromProps(nextProps) {
    return {
      notifications: parseNotifications(nextProps.notifications)
    };
  }

  /**
   * @returns {ParsedNotification[]}
   */
  getNotifications() {
    const {
      dismissedNotifications,
      notifications,
      readNotifications
    } = this.state;

    const updateReadStatus = ({ id, unread, ...otherProps }) => ({
      id,
      unread: unread && !readNotifications.includes(id),
      ...otherProps
    });
    const isNotDismissed = ({ id }) => !dismissedNotifications.includes(id);

    return notifications.map(updateReadStatus).filter(isNotDismissed);
  }

  /** @returns {number} */
  getUnreadCount() {
    const { unreadCount: controlledUnreadCount } = this.props;

    return controlledUnreadCount !== undefined
      ? controlledUnreadCount
      : this.deriveUnreadCount();
  }

  /**
   * Action to dismiss a notification
   * @param {string} id
   */
  dismissNotification = id => {
    this.setState({
      // eslint-disable-next-line react/no-access-state-in-setstate
      dismissedNotifications: this.state.dismissedNotifications.concat(id)
    });
  };

  /**
   * Handler for when the flyout opens
   */
  handleClose = () => {
    this.markAllNotificationsRead();
  };

  /**
   * @param {ParsedNotification[]} notifications
   */
  deriveUnreadCount() {
    return this.getNotifications().reduce(
      (count, { unread }) => (unread ? count + 1 : count),
      0
    );
  }

  markAllNotificationsRead() {
    const notifications = this.getNotifications();
    const { readNotifications } = this.state;
    const nextRead = notifications.reduce((result, notification) => {
      const { id } = notification;

      if (!result.includes(id)) result.push(id);

      return result;
    }, readNotifications.slice());

    this.setState({ readNotifications: nextRead });
  }

  /**
   * @returns {import("react").ReactElement}
   */
  render() {
    const { dismissNotification, handleClose } = this;
    const notifications = this.getNotifications();
    const unreadCount = this.getUnreadCount();
    const showUnreadCount = unreadCount > 0;

    return this.props.children({
      dismissNotification,
      handleClose,
      notifications,
      showUnreadCount,
      unreadCount
    });
  }
}
