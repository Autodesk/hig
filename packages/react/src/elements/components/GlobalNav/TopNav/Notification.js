import React, { Component } from "react";
import PropTypes from "prop-types";

import NotificationAdapter from "../../../../adapters/GlobalNav/TopNav/NotificationAdapter";

class Notification extends Component {
  constructor(props) {
    super(props);
    this.state = {
      unread: this.props.unread
    };
    this.initialProps = props;
  }

  onClick = () => {
    this.setState({ unread: false });
    this.props.onClick(this.props.id);
  };

  render() {
    return (
      <NotificationAdapter
        {...this.props}
        onClick={this.onClick}
        id={this.props.id}
        unread={this.state.unread}
      >
        {this.props.children}
      </NotificationAdapter>
    );
  }
}

Notification.propTypes = {
  unread: PropTypes.bool,
  children: PropTypes.node,
  timestamp: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Date)
  ]), // ISO date string
  onClick: PropTypes.func,
  id: PropTypes.number.isRequired,
  featured: PropTypes.bool
};

Notification.defaultProps = {};

Notification.__docgenInfo = {
  props: {
    id: {
      description: "Id for the notification"
    },
    unread: {
      description: "{Boolean} initializes whether the notification is read"
    },
    children: {
      description: "content for notification"
    },
    timestamp: {
      description: "timstamp for notification"
    },
    onClick: {
      description:
        "Calls the provided callback when user clicks on the noticatiosn icon in the top nav"
    },
    featured: {
      description: "{Boolean} specifies a featured notification"
    }
  }
};

export default Notification;
