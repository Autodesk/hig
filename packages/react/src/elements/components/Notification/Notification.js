import React, { Component } from "react";
import Transition from "react-transition-group/Transition";
import PropTypes from "prop-types";
import cx from "classnames";
import IconButton from "@hig/icon-button";
import Timestamp from "@hig/timestamp";
import RichText from "../RichText";
import "./notification.scss";

const COMPONENT_CLASS = "hig__notification";
const FEATURED_COMPONENT_CLASS = "hig__notification--featured";
const CONTENT_CLASS = "hig__notification__content";

export default class Notification extends Component {
  constructor(props) {
    super(props);
    this.state = {
      in: true
    };
  }

  onClickCapture = event => {
    if (event.target.tagName === "A" && this.props.onLinkClick) {
      this.props.onLinkClick(this.props.id);
    }
  };

  dismissFeaturedNotification = () => {
    this.props.onDismiss();
    this.setState({ in: false });
  };

  render() {
    const notificationClasses = cx(
      `${COMPONENT_CLASS}`,
      { [`hig__notification--${this.props.type}`]: this.props.type },
      {
        [`hig__notification--unread`]: this.props.unread
      }
    );

    const notification = this.props.featured ? (
      <Transition in={this.state.in} timeout={300}>
        {state => (
          <div
            className={`${notificationClasses} hig__notification--${state} ${FEATURED_COMPONENT_CLASS}`}
            onClickCapture={this.onClickCapture}
          >
            <div className={`${CONTENT_CLASS}`}>
              <RichText size="small">{this.props.children}</RichText>
              <IconButton
                onClick={this.dismissFeaturedNotification}
                icon="close-notification"
                title="Dismiss featured notification"
                type="transparent"
              />
            </div>
            {this.props.timestamp ? (
              <Timestamp timestamp={this.props.timestamp} />
            ) : null}
          </div>
        )}
      </Transition>
    ) : (
      <div className={notificationClasses} onClickCapture={this.onClickCapture}>
        <div className={`${CONTENT_CLASS}`}>
          <RichText size="small">{this.props.children}</RichText>
        </div>

        {this.props.timestamp ? (
          <Timestamp timestamp={this.props.timestamp} />
        ) : null}
      </div>
    );

    return notification;
  }
}

Notification.propTypes = {
  /**
   * A callback called when user clicks on a link in the notification
   */
  onLinkClick: PropTypes.func,
  /**
   * A callback called when user dismisses a featured notification
   */
  onDismiss: PropTypes.func,
  /**
   * {Boolean} to show specify whether notificaiton is read
   */
  unread: PropTypes.bool,
  /**
   * Indicates whether this is a featured notification
   */
  featured: PropTypes.bool,
  /**
   * specifies type of notification
   */
  type: PropTypes.string,
  /**
   * id for notification
   */
  id: PropTypes.number,
  /**
   * Content for notification
   */
  children: PropTypes.node,
  /**
   * Timestamp for notification
   */
  timestamp: PropTypes.string // ISO date string
};
