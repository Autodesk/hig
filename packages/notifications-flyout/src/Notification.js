import PropTypes from "prop-types";
import React, { Component } from "react";
import { ControlBehavior } from "@hig/behaviors";

import NotificationBehavior from "./behaviors/NotificationBehavior";
import NotificationPresenter from "./presenters/NotificationPresenter";

export default class Notification extends Component {
  static propTypes = {
    /** Notification content */
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    /** Determines whether the notification is featured */
    featured: PropTypes.bool,
    /** An action provided hide the flyout. This is provided to the `children` render prop */
    hideFlyout: PropTypes.func,
    /** An image to display such as an avatar or and icon */
    image: PropTypes.node,
    /** A callback called when user dismisses a featured notification */
    onDismiss: PropTypes.func,
    /**
     * Triggers when the user's mouse is over the notification
     */
    onMouseEnter: PropTypes.func,
    /**
     * Triggers when the user's mouse is no longer over the notification
     */
    onMouseLeave: PropTypes.func,
    /** Determines whether the dismiss button is shown */
    showDismissButton: PropTypes.bool,
    /** Function to modify the component's styles */
    stylesheet: PropTypes.func,
    /** Timestamp component */
    timestamp: PropTypes.node,
    /** Determines notification variant */
    type: PropTypes.string,
    /** Determines whether notification has not been read */
    unread: PropTypes.bool
  };

  static defaultProps = {
    /**
     * This is an action that's provided to the consumer,
     * as a result a value must always be available.
     */
    hideFlyout: () => {}
  };

  /**
   * @param {NotificationShape} shape
   * @returns {import("react").ReactElement}
   */
  renderChildren() {
    const { children, hideFlyout, onDismiss: dismiss } = this.props;

    if (typeof children !== "function") {
      return children;
    }

    return children({ dismiss, hideFlyout });
  }

  render() {
    const {
      featured,
      image,
      onDismiss,
      onMouseEnter,
      onMouseLeave,
      // Featured notifications show the dismiss button by default
      showDismissButton = featured,
      stylesheet,
      timestamp,
      type,
      unread
    } = this.props;

    return (
      <NotificationBehavior onDismiss={onDismiss}>
        {({ handleDismissButtonClick, height, innerRef, transitionStatus }) => (
          <ControlBehavior
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
          >
            {({
              hasHover,
              onMouseEnter: handleMouseEnter,
              onMouseLeave: handleMouseLeave
            }) => (
              <NotificationPresenter
                featured={featured}
                hasHover={hasHover}
                height={height}
                image={image}
                innerRef={innerRef}
                onDismissButtonClick={handleDismissButtonClick}
                onMouseEnter={handleMouseEnter}
                onMouseLeave={handleMouseLeave}
                showDismissButton={showDismissButton}
                stylesheet={stylesheet}
                timestamp={timestamp}
                transitionStatus={transitionStatus}
                type={type}
                unread={unread}
              >
                {this.renderChildren()}
              </NotificationPresenter>
            )}
          </ControlBehavior>
        )}
      </NotificationBehavior>
    );
  }
}
