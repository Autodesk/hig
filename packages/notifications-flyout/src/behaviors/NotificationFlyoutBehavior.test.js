import React from "react";
import { mount } from "enzyme";

import Notification from "../Notification";
import NotificationFlyoutBehavior from "./NotificationFlyoutBehavior";

describe("notification-flyout/behaviors/NotificationFlyoutBehavior", () => {
  const children = jest.fn();

  function mountBehavior() {
    return mount(
      <NotificationFlyoutBehavior
        notifications={[
          <Notification id="1" key="0" featured>
            Featured
          </Notification>,
          <Notification id="2" key="1">
            Foo
          </Notification>,
          <Notification id="3" key="2">
            Bar
          </Notification>,
        ]}
      >
        {children}
      </NotificationFlyoutBehavior>
    );
  }

  beforeEach(() => {
    children.mockReturnValue(<div />);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders", () => {
    mountBehavior();
  });

  describe("render prop payload", () => {
    let payload;
    let wrapper;

    beforeEach(() => {
      wrapper = mountBehavior();
      [[payload]] = children.mock.calls;
    });

    it("provides a payload to the `children` render prop", () => {
      expect(payload).toMatchObject({
        dismissNotification: expect.any(Function),
        handleClose: expect.any(Function),
        notifications: expect.any(Array),
        showUnreadCount: true,
        unreadCount: 3,
      });
      expect(payload.notifications).toMatchSnapshot();
    });

    it("normalizes notifications", () => {
      expect(payload.notifications).toMatchSnapshot();
    });

    describe("dismissNotification", () => {
      function expectToHaveNotificationTimes(length) {
        const { notifications } = children.mock.calls.slice(-1)[0][0];

        expect(notifications).toHaveLength(length);
      }

      it("dismisses the notification of the given ID", () => {
        const { dismissNotification } = payload;

        expectToHaveNotificationTimes(3);
        dismissNotification("1");
        expectToHaveNotificationTimes(2);
        dismissNotification("1");
        expectToHaveNotificationTimes(2);
      });
    });

    describe("handleClose", () => {
      function expectToHaveUnreadNotificationTimes(length) {
        const { notifications } = children.mock.calls.slice(-1)[0][0];
        const unreadNotifications = notifications.filter(
          ({ unread }) => unread
        );

        expect(unreadNotifications).toHaveLength(length);
      }

      it("marks all notifications as read", () => {
        const { handleClose } = payload;

        expectToHaveUnreadNotificationTimes(3);
        handleClose();
        expectToHaveUnreadNotificationTimes(0);
      });

      it("updates the unread count", () => {
        const { handleClose } = payload;

        expect(children).toHaveBeenLastCalledWith(
          expect.objectContaining({ unreadCount: 3 })
        );

        handleClose();

        expect(children).toHaveBeenLastCalledWith(
          expect.objectContaining({ unreadCount: 0 })
        );
      });
    });

    describe("unreadCount", () => {
      it("is derived from the given notifications", () => {
        expect(payload.unreadCount).toEqual(3);
      });

      describe("when controlled", () => {
        it("it references the `unreadCount` prop", () => {
          const unreadCount = 3;

          wrapper.setProps({ unreadCount });

          expect(children).toHaveBeenLastCalledWith(
            expect.objectContaining({ unreadCount })
          );
        });
      });
    });
  });
});
