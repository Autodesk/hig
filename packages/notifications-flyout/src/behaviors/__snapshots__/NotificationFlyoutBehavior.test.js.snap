// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notification-flyout/behaviors/NotificationFlyoutBehavior render prop payload normalizes notifications 1`] = `
Array [
  Object {
    "content": "Featured",
    "featured": true,
    "hideFlyout": [Function],
    "id": "1",
    "key": "0",
    "onNotificationClick": [Function],
    "unread": true,
  },
  Object {
    "content": "Foo",
    "hideFlyout": [Function],
    "id": "2",
    "key": "1",
    "onNotificationClick": [Function],
    "unread": true,
  },
  Object {
    "content": "Bar",
    "hideFlyout": [Function],
    "id": "3",
    "key": "2",
    "onNotificationClick": [Function],
    "unread": true,
  },
]
`;

exports[`notification-flyout/behaviors/NotificationFlyoutBehavior render prop payload provides a payload to the \`children\` render prop 1`] = `
Array [
  Object {
    "content": "Featured",
    "featured": true,
    "hideFlyout": [Function],
    "id": "1",
    "key": "0",
    "onNotificationClick": [Function],
    "unread": true,
  },
  Object {
    "content": "Foo",
    "hideFlyout": [Function],
    "id": "2",
    "key": "1",
    "onNotificationClick": [Function],
    "unread": true,
  },
  Object {
    "content": "Bar",
    "hideFlyout": [Function],
    "id": "3",
    "key": "2",
    "onNotificationClick": [Function],
    "unread": true,
  },
]
`;
